spec/src/main/asciidoc/query-language.asciidoc

```diff
@@ -0,0 +1,364 @@
+- strictly larger than the value taken by its right operand.
+
+All three operands must have the same type.
+
+==== Like expressions
+
+A `like` expression is satisfied when its left operand evaluates to a non-null value and:
+
+- the `not` keyword is missing, and this value matches the pattern, or
+- the `not` keyword occurs, and the value does not match the pattern.
+
+The left operand must have type `java.lang.String`.
+
+Within the pattern, `_` matches any single character, and `%` matches any sequence of characters.
+
+==== Equality and inequality operators
+
+The equality and inequality operators are `=`, `&lt;&gt;`, `&lt;`, `&gt;`, `&lt;=`, `&gt;=`.
+
+- For primitive types, these operators have the same meaning they have in Java, except for `<>` which has the same meaning that `!=` has in Java. Such an operator expression is satisfied when the equivalent operator expression would evaluate to `true` in Java.
+- For wrapper types, these operators are satisfied if both operands evaluate to non-null values, and the equivalent operator expression involving primitives would be satisfied.
+- For other types, these operators are evaluated according to the native semantics of the database.
+
+NOTE: Portability is maximized when Jakarta Data providers interpret equality and inequality operators in a manner consistent with the implementation of `Object.equals()` or `Comparable.compareTo()` for the assigned Java type.
+
+NOTE: For string values, a database might have a different collation algorithm to Java. In evaluating an inequality involving string operands, an implementation of JDQL is not required to emulate Java collation.
+
+The operands of an equality or inequality operator must have the same type.
+
+==== Logical operators
+
+The logical operators are `and`, `or`, and `not`.
+
+- An `and` operator expression is satisfied if and only if both its operands are satisfied.
+- An `or` operator expression is satisfied if and only if at least one of its operands are satisfied.
+- A `not` operator expression is never satisfied if its operand _is_ satisfied.
+
+This specification leaves undefined the interpretation of the `not` operator when its operand _is not_ satisfied.
+
+CAUTION: A compliant implementation of JDQL might feature SQL/JPQL-style ternary logic, where `not n > 0` is an unsatisfied logical expression when `n` evaluates to null, or it might feature binary logic where the same expression is considered satisfied. Application programmers should take great care when using the `not` operator with scalar expressions involving `null` values.
+
+=== Clauses
+
+Each JDQL statement is built from a sequence of _clauses_. The beginning of a clause is identified by a keyword: `from`, `where`, `select`, `set`, or `order`.
+
+There is a logical ordering of clauses, reflecting the order in which their effect must be computed by the datastore:
+
+1. `from`
+2. `where`,
+3. `select` or `set`,
+4. `order`.
```

Comment:
Are `select` and `order` reversed here?  In the case where `select` is reducing to an entity attribute, it seems like `order` would need to be done first to consider the various other entity attributes to order by.  For example,

`SELECT productId FROM Product ORDER BY price DESC, name ASC, productId ASC`

Reply:
In the full grammar for JPQL we have:

```
select_item ::= select_expression [[AS] result_variable]
select_expression ::=
    single_valued_path_expression |
    scalar_expression |
    aggregate_expression |
    identification_variable |
    OBJECT(identification_variable) |
    constructor_expression

...

orderby_item ::=
    state_field_path_expression |
    general_identification_variable |
    result_variable
    [ASC | DESC]
    [NULLS {FIRST | LAST}]
```

So you may order by any item in the `select` list, including complicated scalar expressions.

So yeah, it's correct.

Reply:
Good point, we need to remain consistent with JPQL even though we don't provide the additional capabilities here